### PR TITLE
Feature/ck 36

### PR DESCRIPTION
**Pull request by Semooze**
Add a renew API

### CODE REVIEW

File: Cpp/odin-views/jwt.renewal.cpp [hunk 1]
```diff
@@ -30,11 +30,11 @@ namespace {
                 fostlib::http::server::request &req,
                 const fostlib::host &host) const {
 
-            if (!req.headers().exists("__app") && !req.headers().exists("__user")) {
+            if (not req.headers().exists("__app") && not req.headers().exists("__user")) {
```
Comment (KayEss) [inline Cpp/odin-views/jwt.renewal.cpp]: Is it not enough to test only `__user` here? Can `__app` app ever be set without `__user`? @tleekkul

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -781,6 +807,7 @@ if(TARGET stress OR TARGET pgtest)
             test-login-google
             test-link-facebook
             test-expire-user
+            jwt-renewal
```
Comment (KayEss) [inline tests/CMakeLists.txt]: It's best to follow the existing naming convention.

File: tests/jwt-renewal.fg [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+# Use APP JWT
+set app_jwt (odin.jwt.mint {"sub": "quser01", "iss": "http://odin.felspar.com/app/app01"} <JWT_SECRET>app01)
+set-path testserver.headers ["Authorization"] (cat "Bearer " (lookup app_jwt))
+
+# Can only use with GET method
+POST test/app-secure/renew-jwt / {} 405
+
+# New JWT is different with the old one
+set new_app_jwt (GET test/app-secure/renew-jwt / 200)
+GET test/not-old-jwt (cat "/" (lookup app_jwt) "/" (lookup new_app_jwt)) 200 {"result" : true}
```
Comment (KayEss) [inline tests/jwt-renewal.fg]: Can't this be done using the `contains` built in?

File: tests/jwt-renewal.fg [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+# New JWT can use without error
+set-path testserver.headers ["Authorization"] (cat "Bearer " (lookup new_app_jwt))
+GET test/app-secure/renew-jwt / 200
+
+# Use non APP JWT
+odin.user user02 user02 password1234
+set jwt (odin.jwt.mint {"sub": "user02"} <JWT_SECRET>)
+set-path testserver.headers ["Authorization"] (cat "Bearer " (lookup jwt))
+PUT test/secure/renew-jwt / {} 405
```
Comment (KayEss) [inline tests/jwt-renewal.fg]: We still need the comments to explain why you're doing this test

File: tests/jwt-renewal.json [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    "odin": {
+        "JWT secret": "<JWT_SECRET>"
+    },
+    "webserver": {
+        "views/test/not-old-jwt": {
+            "view": "fostgres.sql",
+            "configuration": {
+                "host": ["request", "headers", "__pgdsn", "host"],
+                "user": ["request", "headers", "__pgdsn", "user"],
+                "dbname": ["request", "headers", "__pgdsn", "dbname"],
```
Comment (KayEss) [inline tests/jwt-renewal.json]: These three should not be needed now. They're the default.

File: tests/CMakeLists.txt [hunk 6]
```diff
@@ -653,7 +650,6 @@ if(TARGET stress OR TARGET pgtest)
                 ../Schema/opts/installation-id/004-merge-account-function.blue.sql
                 ../Configuration/odin-views.json
                 login-google-test.json
-                facebook-test.json
```
Comment (KayEss) [inline tests/CMakeLists.txt]: Why are all of these being removed?

File: Cpp/odin-views/jwt.renewal.cpp [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+#include <odin/odin.hpp>
+#include <odin/user.hpp>
+#include <odin/views.hpp>
+
+#include <fostgres/sql.hpp>
+
+#include <fost/insert>
+#include <fost/json>
+#include <fost/push_back>
+#include <fost/string.hpp>
```
Comment (KayEss) [inline Cpp/odin-views/jwt.renewal.cpp]: This looks wrong. The headers you are meant to use don't end in `.hpp`

File: Cpp/odin-views/jwt.renewal.cpp (hunk 1 above)
Comment (tleekkul) [reply Cpp/odin-views/jwt.renewal.cpp]: `__app` can be set without `__user`.

Comment (KayEss) [reply Cpp/odin-views/jwt.renewal.cpp]: Interesting. What does that mean? App based JWT with no logged in user? Then it should go down the insecure path for the secure view?

Comment (tleekkul) [reply Cpp/odin-views/jwt.renewal.cpp]: It means this is the app token, not user token. Maybe we want to change this behavior to be every jwt have to have `sub` value. For the app token, `sub` must refer to the app ID.

Comment (KayEss) [reply Cpp/odin-views/jwt.renewal.cpp]: I think my point is that this particular API can only be for refreshing the JWT for a logged in user. It cannot be used for anything else. The app JWT for when a user is not logged in cannot expire I think or we have other problems. Doesn't that affect which of these can be set?

Comment (tleekkul) [reply Cpp/odin-views/jwt.renewal.cpp]: I get your point. Yes, it enough to test only `__user` here.

Comment (KayEss) [reply Cpp/odin-views/jwt.renewal.cpp]: And we don't set expiry on the app JWT do we? What's the expectation for what should happen once the log in JWT expires? Got to say I think I'm still somewhat confused about how some of this meant to work. I think we should go through on the whiteboard and then write it all down.

**Pull request merged**